File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:

        if (head.next == None):
            return None
        
        node = head
        tmp = head

        i=1
        while (i<=n):
            node = node.next
            i+=1
        
        if (node == None):
            return head.next

        while(node.next !=None):
            node = node.next
            tmp = tmp.next

        # if (tmp.next!=None):
        tmp.next = tmp.next.next

        return head

"""
Ý tưởng: 
- cho node đi trước n bước để tạo chênh lệch với tmp
- Sau đó cho node và tmp cùng đi:
 => Nếu node gặp node cuối nghĩa là -> tmp cách end đúng n nodes.
"""

File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        int sz = 0;

        ListNode* cur = head;
        while(cur){
            sz++;
            cur = cur->next;
        }
        if (sz==1)
            return nullptr;
        if (n==sz)
            head = head->next;

        int i=0;
        cur = head;

        while(cur){
            if (i==(sz-1)-n){
                cur->next = cur->next->next;
            }
            i++;
            cur=cur->next;
        }

        cout<<sz;

        return head;
    }
};